examples/vision/metric_learning.py

```diff
@@ -22,290 +24,380 @@
+select a block of samples.
 """
 
-class_idx_to_train_idxs = defaultdict(list)
-for y_train_idx, y in enumerate(y_train):
-    class_idx_to_train_idxs[y].append(y_train_idx)
+num_cols = num_rows = 5
+# Get the first 25 examples.
+X, y = train_ds.get_slice(begin=0, size=num_cols * num_rows)
 
-class_idx_to_test_idxs = defaultdict(list)
-for y_test_idx, y in enumerate(y_test):
-    class_idx_to_test_idxs[y].append(y_test_idx)
+fig = plt.figure(figsize=(6., 6.))
+grid = ImageGrid(fig, 111, nrows_ncols=(num_cols, num_rows), axes_pad=0.1)
+
+for ax, im, label in zip(grid, X, y):
+    ax.imshow(im)
+    ax.axis('off')
 
 """
-For this example we are using the simplest approach to training; a batch will consist of
-`(anchor, positive)` pairs spread across the classes. The goal of learning will be to
-move the anchor and positive pairs closer together and further away from other instances
-in the batch. In this case the batch size will be dictated by the number of classes; for
-CIFAR-10 this is 10.
+## Batch Sizes
+
+Additionally, we can use the `generate_batch()` method to yield a batch and
+check that it contains the expected number of classes and examples per class.
 """
 
-num_classes = 10
+x_batch, y_batch = train_ds.generate_batch(0)
+unique_class_ids, _, class_counts = tf.unique_with_counts(y_batch)
+print(f'The batch contains {len(x_batch)} examples in total\n')
+for cid, cc in zip(unique_class_ids, class_counts):
+    print(f'Class ID {cid} has {cc} examples in the batch')
 
+"""
+# Embedding Model
 
-class AnchorPositivePairs(keras.utils.Sequence):
-    def __init__(self, num_batchs):
-        self.num_batchs = num_batchs
+Next we define a `SimilarityModel()` using the functional Keras API. The model
+is a standard ConvNet with the addition of a `MetricEmbedding()` layer that
+applies L2 normalization. The metric embedding layer is helpful when using
+`Cosine` distance as we only care about the angle between the vectors.
 
-    def __len__(self):
-        return self.num_batchs
+Additionally, the `SimilarityModel()` provides a number of helper methods for:
 
-    def __getitem__(self, _idx):
-        x = np.empty((2, num_classes, height_width, height_width, 3), dtype=np.float32)
-        for class_idx in range(num_classes):
-            examples_for_class = class_idx_to_train_idxs[class_idx]
-            anchor_idx = random.choice(examples_for_class)
-            positive_idx = random.choice(examples_for_class)
-            while positive_idx == anchor_idx:
-                positive_idx = random.choice(examples_for_class)
-            x[0, class_idx] = x_train[anchor_idx]
-            x[1, class_idx] = x_train[positive_idx]
-        return x
+* indexing embedded examples
+* performing example lookups
+* evaluating the classification
+* evaluating the quality of the embedding space
+"""
 
+embedding_size = 256
+
+inputs = layers.Input((32, 32, 3))
+x = inputs
+x = layers.Conv2D(64, 3, activation='relu')(x)
```

Comment:
Add the normalization and a simple aug layers.

Reply:
Added the normalization but we ended up omitting the augmentation layers here as they didn't seem to improve the loss here.